### ISSUE TITLE
Suggestion for new rule: case where a function passed a callback parameter does not call it.

### ISSUE THREAD

**Issue by autopulated** [eslint/eslint#9351]
Apologies if this is unwelcome, but in reviewing some code I noticed a class of bug that I think eslint could probably detect, so I wanted to suggest it for a new rule:

**Please describe what the rule should do:**
If a function is passed a callback parameter, warn if there's a case where the function would return without calling the supplied callback or passing it to another function.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
function buggyFn(a, b, callback){
    if(a){
       doSomething(callback)
    }else if(b){
        doSomethingElse(); // likely a bug here, callback goes out of scope without being called
    }else{
        return callback();
    }
}

function anotherBuggyFn(cb){
  if(something()){
    return; // likely bug here, in this case cb() will not be called before function return
  }
  doSomeWork();
  return cb();
}

function anotherFn(a, b, c, cb){
  a.foo(function(err){
    if(err) return; // likely a bug here, callback goes out of scope without being called
    b.bar(function(err){
      if(err) return cb(err);
      else return c.baz(cb);
    })
  });
}

```

**Why should this rule be included in ESLint (instead of a plugin)?**
Widely applicable category of programming error.

**Comment by ilyavolodin** [comment]
I guess this would fit well with `handle-callback-err` and `callback-return`. So I'll support this. @autopulated Could you update first post with the new rule proposal template please? https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md

**Comment by autopulated** [comment]
done  :)

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
